android/app/src/main/kotlin/com/zulip/flutter/MainActivity.kt

```diff
@@ -1,6 +1,131 @@
 package com.zulip.flutter
 
+import android.content.Intent
+import android.net.Uri
+import android.provider.OpenableColumns
 import io.flutter.embedding.android.FlutterActivity
+import io.flutter.embedding.engine.FlutterEngine
 
-class MainActivity: FlutterActivity() {
+class MainActivity : FlutterActivity() {
+  private var androidIntentEventListener: AndroidIntentEventListener? = null
+
+  override fun configureFlutterEngine(flutterEngine: FlutterEngine) {
+    super.configureFlutterEngine(flutterEngine)
+
+    androidIntentEventListener = AndroidIntentEventListener()
+    AndroidIntentEventsStreamHandler.register(
+      flutterEngine.dartExecutor.binaryMessenger,
+      androidIntentEventListener!!
+    )
+    maybeHandleIntent(intent)
+  }
+
+  override fun onNewIntent(intent: Intent) {
+    if (maybeHandleIntent(intent)) {
+      return
+    }
+    super.onNewIntent(intent)
+  }
+
+  /** Returns true just if we did handle the intent. */
+  private fun maybeHandleIntent(intent: Intent?): Boolean {
+    intent ?: return false
+    when (intent.action) {
+      // Share-to-Zulip
+      Intent.ACTION_SEND, Intent.ACTION_SEND_MULTIPLE -> {
+        handleSend(intent)
+        return true
+      }
+
+      // For other intents, let Flutter handle it.
+      else -> return false
+    }
+  }
+
+  private fun getIntentSharedFileInfo(url: Uri) : IntentSharedFile {
+    val mimeType = contentResolver.getType(url)
+    var name = contentResolver.query(url, null, null, null, null)?.use { cursor ->
+      cursor.moveToFirst()
+      val nameIndex = cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME)
+      cursor.getString(nameIndex)
+    } ?: "unknown." + (mimeType?.split('/')?.last() ?: "bin")
+
+    class ResolverFailedException(msg: String) : RuntimeException(msg)
+
+    val bytes = (contentResolver.openInputStream(url)
+      ?: throw ResolverFailedException("resolver.open… failed"))
+      .use { inputStream ->
+        inputStream.readBytes()
+      }
+
+    return IntentSharedFile(
+      name = name,
+      mimeType = mimeType,
+      bytes = bytes)
+  }
+
+  private fun handleSend(intent: Intent) {
```

Comment:
Can this method (and its helper getIntentSharedFileInfo) move onto a class other than MainActivity? Perhaps on AndroidIntentEventListener, since that seems closely related.

That would help for making it explicit what information or methods they're using, if any, from FlutterActivity or its base classes.

Even if the methods have to be passed `this` as a FlutterActivity (or some supertype of that), I think the explicitness of accessing FlutterActivity members on that argument would be helpful.